File: src/traust_engine/data/secure-code-audit/opengrep-rules/python/injection.py
import os
import subprocess

from flask import request


def run_dynamic(user_target):
    # ruleid: traust-python-injection-subprocess-shell
    subprocess.run(f"nslookup {user_target}", shell=True)

    # ruleid: traust-python-injection-subprocess-shell
    os.system("ping -c1 " + user_target)

    # ok: traust-python-injection-subprocess-shell
    subprocess.run("systemctl restart chronyd", shell=True)

    # ok: traust-python-injection-subprocess-shell
    subprocess.run(["nslookup", user_target])


def handler():
    host = request.args.get("host")
    # ruleid: traust-python-injection-request-to-shell-taint
    subprocess.check_output(["dig", host])

    # ok: traust-python-injection-request-to-shell-taint
    subprocess.check_output(["dig", "example.com"])


def raw_sql(request, connection, Model):
    q = request.GET.get("q")
    cursor = connection.cursor()
    # ruleid: traust-python-injection-raw-sql-taint
    cursor.execute("SELECT * FROM t WHERE name = '%s'" % q)

    # ruleid: traust-python-injection-raw-sql-taint
    Model.objects.extra(where=["name = '%s'" % q])

    # ok: traust-python-injection-raw-sql-taint
    cursor.execute("SELECT * FROM t WHERE fixed = 1")


def render_user_template(request):
    import jinja2
    tpl = request.form.get("template")
    # ruleid: traust-python-injection-jinja2-template-taint
    return jinja2.Template(tpl).render()


def render_static_template():
    import jinja2
    # ok: traust-python-injection-jinja2-template-taint
    return jinja2.Template("hello {{ name }}").render(name="world")

File: src/traust_engine/data/secure-code-audit/opengrep-rules/java/injection.java
import java.sql.Connection;
import java.sql.Statement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import javax.persistence.EntityManager;

class InjectionTests {

    void sql(Statement stmt, Connection conn, EntityManager em, String user) throws Exception {
        // ruleid: traust-java-injection-sql-concat
        ResultSet rs = stmt.executeQuery("SELECT * FROM users WHERE name = '" + user + "'");
        // ruleid: traust-java-injection-sql-concat
        stmt.executeUpdate("DELETE FROM sessions WHERE owner = '" + user + "'");
        // ruleid: traust-java-injection-sql-concat
        stmt.execute("UPDATE t SET v = 1 WHERE k = '" + user);
        // ruleid: traust-java-injection-sql-concat
        PreparedStatement ps = conn.prepareStatement("SELECT * FROM users WHERE name = '" + user + "'");
        // ruleid: traust-java-injection-sql-concat
        em.createNativeQuery("SELECT * FROM users WHERE name = '" + user + "'");

        // ok: traust-java-injection-sql-concat
        PreparedStatement safe = conn.prepareStatement("SELECT * FROM users WHERE name = ?");
        safe.setString(1, user);
        // ok: traust-java-injection-sql-concat
        ResultSet fixed = stmt.executeQuery("SELECT COUNT(*) FROM users");
    }

    void exec(String host) throws Exception {
        // ruleid: traust-java-injection-exec-concat
        Runtime.getRuntime().exec("ping -c 1 " + host);
        Runtime rt = Runtime.getRuntime();
        // ruleid: traust-java-injection-exec-concat
        rt.exec("nslookup " + host + " 8.8.8.8");
        // ruleid: traust-java-injection-exec-concat
        new ProcessBuilder("sh", "-c", "dig " + host);

        // ok: traust-java-injection-exec-concat
        Runtime.getRuntime().exec(new String[]{"ping", "-c", "1", host});
        // ok: traust-java-injection-exec-concat
        new ProcessBuilder("ping", "-c", "1", host);
    }
}
